### PR TITLE
feat(common/core/web): fat-finger ignores inputs that beep

### LINKED ISSUE #2025: Use of keyboard-rule negative feedback in fat-finger probability generation
From a review comment on #2013:

> Still amused by this one... The `beep` effect could be a way of providing negative feedback to fat finger algorithm too: it's unlikely the user accidentally wanted to type the invalid sequence, right?

It would make sense to use this when weighting the likelihood of fat-finger keypresses.

### PR DESCRIPTION
**Pull request by jahorton**
Fixes #2025.

Turns out most of the KMW refactor work from earlier in 14.0 made fixing this issue really easy.

~~That said... I don't know of a good test case for thorough testing; we'd need a language with a lexical model that also has a touch keyboard with outputs that beep.~~

### CODE REVIEW

**Comment by mcdurdin** [comment]
> That said... I don't know of a good test case for thorough testing; we'd need a language with a lexical model that also has a touch keyboard with outputs that beep.

Easy enough to build a keyboard that triggers this though?

**Comment by mcdurdin** [comment]
[test3701.zip](https://github.com/keymanapp/keyman/files/5398454/test3701.zip)

This should do the job I hope?

File: web/testing/issue3701/test3701/source/welcome.htm [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    h1 { font: bold 16pt Tahoma; color: #4444cc; margin-bottom: 2px }
+    h2 { font: bold 12pt Tahoma; color: #4444cc; }
+  </style>
+</head>
+<body>
+
+<h1>Start Using test3701</h1>
+
+<p>
+    test3701 1.0 generated from template.
```
Comment (darcywong00) [inline web/testing/issue3701/test3701/source/welcome.htm:18]: Even though this is from template, maybe add a description that the keyboard prohibits use of the letter <kbd>x</kbd> (and <kbd>X</kbd>)?

**Comment by jahorton** [comment]
Good thing we enhanced the test keyboard and went another round of testing; that caught a significant bug.

With that fixed and out of the way... a minor demo yields this:

Screen Shot 2020-10-20 at 9 57 14 AM

As all `x` keystrokes emit 'beeps' and are thus interpreted as errors, predictive text now tends to avoid suggestions with 'x' (as they require a low-probability 'substitution', even if that 'x' is the actual incoming keystroke.

Screen Shot 2020-10-20 at 10 03 06 AM

It actually tends to prevent _any_ suggestions with an 'x' from appearing... probably due to some of the other thresholding settings.  I might should investigate that a bit.

**Comment by jahorton** [comment]
Further investigation suggests that my observation was due to choosing examples with 'x' early in the word - a scenario in which many, _many_ other corrections are possible and were considered "more likely."  Performance-related thresholding prevented us from reaching those suggestions.

Choosing a word with 'x' at a later position, once the possibilities are far more constrained, has a better chance of working:

Screen Shot 2020-10-20 at 12 06 35 PM

"ambidextrous" remains even after reaching and typing the 'beeped' `x`.

Compare against:

Screen Shot 2020-10-20 at 12 07 38 PM

At no point did "dexterity" display as an option.  It doesn't help that "des" and "dest" prefix a lot of English words, with 's' neighboring the 'x' key.

**Comment by mcdurdin** [comment]
As I reviewed your examples, I realised there's a question in my mind as to the practical reality of this. Surely if it is an invalid sequence, it won't appear in the dictionary anyway? So 'corrections' that traversed that key would never be offered, because they wouldn't exist... and so they'd automatically fall out. The 'x' example is confected because it doesn't match how a real keyboard works!

**Comment by jahorton** [comment]
There are some indirect benefits.  If you examine this alongside #3726, note that the two PRs remove a number of keys from the fat-finger distribution... increasing the "probability mass" of the remaining keystrokes.

Since we _do_ apply thresholding to the overall probability of a keystroke sequence within the predictive-text engine, these changes can allow some 'correction' sequences to pass the threshold when they previously would have failed.  The smaller distributions also yield some _slight_ memory + time savings, as the useless search paths currently consume some computational resources before they become ignored by the search algorithm.

File: common/core/web/input-processor/src/text/inputProcessor.ts [hunk 2]
```diff
@@ -128,19 +129,25 @@ namespace com.keyman.text {
...
                 transform.id = ruleBehavior.transcription.token;
                 alternates.push({sample: transform, 'p': pair.p});
+                totalMass += pair.p;
               }
             }
+
+            // Renormalizes the distribution, as any error (beep) results
+            // will result in a distribution that doesn't sum to 1 otherwise.
+            alternates.forEach(function(alt) {
+              alt.p /= totalMass;
```
Comment (mcdurdin) [inline common/core/web/input-processor/src/text/inputProcessor.ts:149]: Is there ever a possibility of totalMass == 0?

**Review by mcdurdin — APPROVED** [review]
Okay, the change is simple enough, and clear.

Apart from that single question, LGTM.

File: common/core/web/input-processor/src/text/inputProcessor.ts (hunk 2 above)
Comment (jahorton) [reply common/core/web/input-processor/src/text/inputProcessor.ts:149]: It's the same probability of a keyboard being designed that could conceivably output a BEEP for literally every output keystroke on one layer for the same context.

So... I'm going with "practically, no."  It should take very, very poor design for that to happen.

Comment (mcdurdin) [reply common/core/web/input-processor/src/text/inputProcessor.ts:149]: > It should take very, very poor design for that to happen.

So, practically guaranteed to happen at some point then? 🤣

**Pull request merged**